_scripts/utils.py

```diff
@@ -0,0 +1,303 @@
+
+ALLOWED_TYPES = {'article', 'book', 'conference', 'inbook', 'phdthesis', 'techreport', 'tutorial', 'unpublished'}
+
+ALLOWED_ICONS = {
+    'article-line', 'book-line', 'braces-line', 'close-circle-fill',
+    'close-circle-line', 'close-fill', 'close-line', 'code-line',
+    'double-quotes-r', 'file-line', 'file-text-line', 'git-repository-line',
+    'image-line', 'link', 'mail-line', 'profile-line',
+    'slideshow-line', 'video-line', 'video',
+}
+
+REQUIRED_KEYS = [
+    'type',
+    'authors',
+    'title',
+    'year',
+    'venue',
+    'publisher',
+]
+
+# Keys (outside of REQUIRED_KEYS) that are required for specific types.
+TYPE_FIELDS = {
+    'article': {'number', 'pages', 'volume'},
+    'book': {'edition'},
+    'conference': {'address'},
+    'inbook': {'edition', 'editor', 'pages'},
+    'phdthesis': {'address'},
+    'techreport': set(),
+    'tutorial': {'address'},
+    'unpublished': set(),
+}
+
+# Also doubles as the allowed keys.
+KEY_TYPES = {
+    'abstract': [str],
+    'address': [str],
+    'authors': [list],
+    'chapter': [int, str],
+    'doi': [str],
+    'edition': [int, str],
+    'editor': [str],
+    'links': [list],
+    'keywords': [list],
+    'note': [str],
+    'number': [int, str],
+    'organization': [str],
+    'pages': [int, str],
+    'publisher': [str],
+    'title': [str],
+    'type': [str],
+    'venue': [str],
+    'volume': [int, str],
+    'year': [int, str],
+}
+
+# Special Venue IDs that depend on the pub's type instead of the pub's venue.
+TYPE_VENUE_IDS = {
+    'book': 'book',
+    'phdthesis': 'phd',
+    'techreport': 'techreport',
+}
+
+def validateOptionalNestedKeys(filename, data):
+    errors = []
+
+    for (optional_key, optional_value) in OPTIONAL_NESTED_KEY_TYPES.items():
+        if (optional_key not in data):
+            continue
+
+        else:
+            for nested_data in data[optional_key]:
+                keys_present = set()
+
+                for (nested_key, nested_value) in optional_value.items():
+                    if (nested_key in nested_data):
+                        keys_present.add(nested_key)
+                        if (type(nested_data[nested_key]) not in optional_value[nested_key]):
+                            errors.append("Incorrect type ('%s') found in %s." % (nested_key, filename))
+                            continue
+
+                num_keys = len(optional_value.items())
+                keys_absent = set(optional_value.keys()).difference(keys_present)
+
+                if (len(keys_present) != 0 and len(keys_present) != num_keys):
+                    errors.append("Optional keys must either be all included or all excluded. Found %s but not %s in %s." % (", ".join(keys_present), ",".join(keys_absent), filename))
+
+    return errors
+
+def validateRequiredNestedKeys(filename, data):
+    errors = []
+
+    for (required_key, required_value) in REQUIRED_NESTED_KEY_TYPES.items():
+        if (required_key in data):
```

Comment:
Do you mean invert them logically?

Reply:
Yeah.
Instead of having this long running conditional, you can just check for the inversion and bail out.
Something like:
```
it (key not in data):
    continue
```

Remember, every time you create a new scope, a developer has to remember the conditions of that scope.
Doing things like this reduces the mental burden on developers (and simplifies code).